File: java/trip-service-kata/src/test/java/org/craftedsw/tripservicekata/trip/TripServiceShould.java
package org.craftedsw.tripservicekata.trip;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;
import java.util.List;

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.User;


public class TripServiceShould {
	private static final User REGISTERED_USER = new User();
	private static final User ANOTHER_USER = new User();
	private static final Trip BARCELONA = new Trip();
	private User loggedInUser;
	private TripService tripService;

	@Before
	public void initialize(){
		tripService = new TestableTripService();
	}
	
	@Test(expected=UserNotLoggedInException.class)
	public void validate_logged_in_user() {		
		User user = new User();		
		tripService.getTripsByUser(user);
	}
			
	@Test
	public void return_no_trips_when_logged_in_user_has_no_friends(){
		this.loggedInUser = REGISTERED_USER;
		
		User friend = new User();
		friend.addFriend(ANOTHER_USER);
		friend.addTrip(BARCELONA);
		
		
		List<Trip> trips = tripService.getTripsByUser(friend);
		assertThat(trips.size(), is(0));
	}
	
	@Test
	public void return_trips_when_users_are_friends() {
		this.loggedInUser = REGISTERED_USER;
		
		User user = new User();
		user.addFriend(REGISTERED_USER);
		user.addFriend(ANOTHER_USER);
		user.addTrip(BARCELONA);
		
		List<Trip> trips = tripService.getTripsByUser(user);
		assertThat(trips.size(), is(1));
		
	}	

	private class TestableTripService extends TripService{

		@Override
		protected User getLoggedUser() {
			return loggedInUser;
		}
		
		@Override
		protected List<Trip> findTripsBy(User user) {
			return user.trips();
		}		
	}
}

File: src/test/kotlin/tripservicekata/trip/TripServiceShould.kt
package org.craftedsw.tripservicekata.trip

import org.assertj.core.api.Assertions
import org.craftedsw.tripservicekata.exception.UserNotLoggedInException
import org.craftedsw.tripservicekata.user.User
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.ArrayList

class TripServiceShould {

    @Test
    fun `throw and exception when user is not logged in`() {
        val anonymousUser = null
        val user = User()

        val tripService = TripServiceMock(anonymousUser)

        assertThrows<UserNotLoggedInException> {
            tripService.getTripsByUser(user)
        }
    }

    @Test
    fun `not return trips when the user is logged in and has no friends`() {
        val registeredUser = User()
        val user = User()
        val tripService = TripServiceMock(registeredUser)

        val trips = tripService.getTripsByUser(user)

        Assertions.assertThat(trips).isEqualTo(ArrayList<Trip>())
    }

    class TripServiceMock(private val user: User?) : TripService() {

        override fun getLoggedUser(): User? {
            return user
        }
    }
}

